jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
     }
 
-    private static class SubscriptionImpl implements Flow.Subscription
+    private void onSubscribe(Flow.Subscriber<? super Content.Chunk> subscriber, Content.Source content)
     {
-        private final AutoLock lock = new AutoLock();
-        private final Content.Source content;
-        private final Flow.Subscriber<? super Content.Chunk> subscriber;
-        private long demand;
-        private boolean stalled;
-        private boolean cancelled;
-        private boolean terminated;
+        // As per rule 1.9, we need to throw a `java.lang.NullPointerException`
+        // if the `Subscriber` is `null`
+        if (subscriber == null)
+        {
+            NullPointerException error = new NullPointerException("Flow.Subscriber must not be null");
+            content.fail(error);
+            throw error;
+        }
+        LastWillSubscription subscription = new ActiveSubscription(content, subscriber);
+        // As per rule 1.9, this method must return normally (i.e. not throw).
+        try
+        {
+            subscriber.onSubscribe(subscription);
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+        }
+    }
 
-        public SubscriptionImpl(Content.Source content, Flow.Subscriber<? super Content.Chunk> subscriber)
+    private void onMultiSubscribe(Flow.Subscriber<? super Content.Chunk> subscriber)
+    {
+        // As per rule 1.9, we need to throw a `java.lang.NullPointerException`
+        // if the `Subscriber` is `null`
+        if (subscriber == null)
         {
-            this.content = content;
-            this.subscriber = subscriber;
-            this.stalled = true;
+            throw new NullPointerException("Flow.Subscriber must not be null");
         }
+        LastWillSubscription subscription = new ExhaustedSubscription();
 
+        // As per 1.9, this method must return normally (i.e. not throw).
+        try
+        {
+            // As per rule 1.9, the only legal way to signal about Subscriber rejection
+            // is by calling onError (after calling onSubscribe).
+            subscriber.onSubscribe(subscription);
+            subscriber.onError(new IllegalStateException("Content.Source was exhausted."));
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+        }
+    }
```

Comment:
I think we need to try/catch the `onSubscribe` and `onError` separately:
```suggestion
        // As per 1.9, this method must return normally (i.e. not throw).
        Throwable exhausted = new IllegalStateException("Content.Source was exhausted.");
        try
        {
            // As per rule 1.9, the only legal way to signal about Subscriber rejection
            // is by onSubscribe and then onError
            subscriber.onSubscribe(subscription);
        }
        catch (Throwable err)
        {
            ExceptionUtil.addSuppressedIfNotAssociated(exhausted, err);
        }
        
        try
        {
            // As per rule 1.9, the only legal way to signal about Subscriber rejection
            // is by calling onError (after calling onSubscribe).
            subscriber.onError(exhausted);
        }
        catch (Throwable t)
        {
            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", t);
        }
```

Reply:
@gregw I had clarified in another comment, but the point was
- Subscriber.onSubscribe/onNext/onError/onComplete failed - also must be suppressed according to the specification. It means, that if Subscriber.onSubscribe failed you must not execute Subscriber.onError, because that Subscriber considering as broken and any further methods invocation on the subscriber may not work as well